File: app/src/main/java/io/github/pshegger/infobook/adapters/CharacterListAdapter.java
package io.github.pshegger.infobook.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import io.github.pshegger.infobook.R;
import io.github.pshegger.infobook.model.CharacterData;
import io.github.pshegger.infobook.utils.IconLoadHelper;
import io.github.pshegger.infobook.utils.OnRecyclerViewItemClickListener;

/**
 * Created by pshegger on 2015.10.13..
 */
public class CharacterListAdapter extends RecyclerView.Adapter<CharacterListAdapter.ViewHolder> {
    private List<CharacterData> mCharacters;
    private OnRecyclerViewItemClickListener mListener;

    public CharacterListAdapter(List<CharacterData> characters, OnRecyclerViewItemClickListener listener) {
        mCharacters = characters;
        mListener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.adapter_character_row, parent, false);

        return new ViewHolder(v, mListener);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(getItem(position), position);
    }

    @Override
    public int getItemCount() {
        return mCharacters.size();
    }

    public CharacterData getItem(int position) {
        return mCharacters.get(position);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private ImageView ivLogo;
        private TextView tvName, tvProfession, tvLevel;
        private View root;

        private OnRecyclerViewItemClickListener mListener;
        private int mPosition;

        public ViewHolder(View itemView, OnRecyclerViewItemClickListener listener) {
            super(itemView);

            root = itemView;
            mListener = listener;

            ivLogo = (ImageView) itemView.findViewById(R.id.profession_logo);
            tvName = (TextView) itemView.findViewById(R.id.character_name);
            tvProfession = (TextView) itemView.findViewById(R.id.character_profession);
            tvLevel = (TextView) itemView.findViewById(R.id.character_level);

            if (listener != null) {
                root.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mListener.onItemClick(mPosition);
                    }
                });
            }
        }

        public void bind(CharacterData ch, int position) {
            mPosition = position;

            tvName.setText(ch.getName());
            tvProfession.setText(ch.getRace().name());
            tvLevel.setText(String.valueOf(ch.getLevel()));

            IconLoadHelper.loadIcon(
                    IconLoadHelper.IconType.File,
                    ch.getProfession().getIconId(true),
                    ivLogo
            );
        }
    }
}

File: app/src/main/kotlin/io/github/pshegger/infobook/adapters/CharacterListAdapter.kt
package io.github.pshegger.infobook.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import io.github.pshegger.infobook.R
import io.github.pshegger.infobook.model.CharacterData
import io.github.pshegger.infobook.utils.IconLoadHelper
import io.github.pshegger.infobook.utils.OnRecyclerViewItemClickListener

/**
 * Created by pshegger on 2015. 11. 10..
 */
class CharacterListAdapter(val characters: List<CharacterData>, val listener: OnRecyclerViewItemClickListener?) : RecyclerView.Adapter<CharacterViewHolder>() {
    override fun onBindViewHolder(holder: CharacterViewHolder, position: Int) {
        holder.bind(this[position], position)
    }

    override fun getItemCount(): Int = characters.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CharacterViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.adapter_character_row, parent, false)

        return CharacterViewHolder(v, listener)
    }

    operator fun get(position: Int) = characters[position]
}

class CharacterViewHolder(val root: View, val listener: OnRecyclerViewItemClickListener?) : RecyclerView.ViewHolder(root) {
    val ivLogo = root.findViewById(R.id.profession_logo) as ImageView
    val tvName = root.findViewById(R.id.character_name) as TextView
    val tvProfession = root.findViewById(R.id.character_profession) as TextView
    val tvLevel = root.findViewById(R.id.character_level) as TextView
    var mPosition = -1

    init {
        listener?.let {
            root.setOnClickListener { listener.onItemClick(mPosition) }
        }
    }

    fun bind(ch: CharacterData, position: Int) {
        mPosition = position

        tvName.text = ch.name
        tvProfession.text = ch.profession().name
        tvLevel.text = "" + ch.level

        IconLoadHelper.loadIcon(
                IconLoadHelper.IconType.File,
                ch.profession().iconId(true),
                ivLogo
        )
    }
}